### PR TITLE
Fix #47 and #70

### LINKED ISSUE #47: Refactor tests
Refactor test suit.

Based on https://github.com/BurntSushi/toml-test

### PR DESCRIPTION
**Pull request by timcolonel**
I added a few features trying to fix #47 and #70.
* Subclasses the 2 error classes with `TOML::Errror` (#70)
* Added a list of valid file to be parsed with toml and the json equivalent. Unit test is testing against each file.
* This lead to a few unsupported feature:
  - Unicode escape characters(\u03B4)
  - Empty multiline expression(`"""""""`)
* I also replaced Unit::Test with Minitest::Test which i think is now recommended over Unit::Test
* The error `ValueOverwriteError` was too strict. The following was failing to parse but is suppose to work:
```toml
[a.b.c]
answer = 42

[a]
better = 43
```
* Having spaces in keys withour quotes is supposed to raise an error `[invalid key]` `["valid key"]` so modified the grammar to be more strict. (check `test/examples/invalid/table-whitespace.toml`)

If anything seems wrong, tell me and I will fix it.

### CODE REVIEW

File: .gems [hunk 1]
```diff
@@ -1 +1,2 @@
 citrus -v 3.0.2
+minitest -v 5.7.0
```
Comment (emancu) [inline .gems]: Why do you say that `minitest` is better ?
Honestly I didn't use `minitest` because I wanted to keep simple the `gems dependencies` even for dev.

Anyways, it's ok, I just wonder why did you need it?

File: test/grammar_test.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # encoding: utf-8
 require_relative 'helper'
 
-class GrammarTest < Test::Unit::TestCase
+class GrammarTest < Minitest::Test
```
Comment (emancu) [inline test/grammar_test.rb]: Maybe we can remove the `GrammarTest` if we have a complete suite as you did.
For example, the empty multiline string, could be a file `/examples/valid/empty-multiline-strings.toml`.

What do you think?

File: .gems (hunk 1 above)
Comment (timcolonel) [inline .gems]: If you look on the unit::test documentation http://ruby-doc.org/stdlib-2.1.1/libdoc/test/unit/rdoc/Test/Unit.html it state `If you are writing new test code, please use MiniTest instead of Test::Unit.`

File: test/toml_test.rb [hunk 3]
```diff
@@ -84,4 +85,34 @@ def test_line_break
...
+      rescue TOML::Error => e
+        assert false, "Error: #{e} in #{toml_file}"
+      end
+      json = JSON.parse(File.read(json_file))
+      block.call(json, toml, toml_file)
+    end
+  end
+
+  def test_valid_cases
+    compare_toml_files 'valid', 'table-whitespace' do |json, toml, file|
```
Comment (emancu) [inline test/toml_test.rb]: You are testing only one file here.

File: test/grammar_test.rb (hunk 2 above)
Comment (timcolonel) [inline test/grammar_test.rb]: I think this could be a good idea. However as currently i am just comparing the toml hash and json hash having complex toml files makes debuging hard.

File: test/toml_test.rb [hunk 4]
```diff
@@ -84,4 +85,34 @@ def test_line_break
     parsed = TOML.parse("hello = 'world'\r\nline_break = true")
     assert_equal({ 'hello' => 'world', 'line_break' => true }, parsed)
   end
+
+  def compare_toml_files(folder, file = nil, &block)
```
Comment (emancu) [inline test/toml_test.rb]: This is a great method, only used by `test_valid_cases` but not for `test_invalid_cases`. Is there any reason to do that?

Also, this should be a `private` I guess.

File: test/toml_test.rb (hunk 3 above)
Comment (timcolonel) [inline test/toml_test.rb]: Going to fix that

File: .gems (hunk 1 above)
Comment (emancu) [inline .gems]: You have a good point. :+1:

**Comment by emancu** [comment]
Awesome PR!!

Thank you very much for taking care of those issues, I was a little busy to do it.

**Comment by timcolonel** [comment]
On the other hand keeping `test_grammar` allow to test more specific case. Also I could not test the datetime parsing automaticaly with json comparing as json doesn't support it so we need to keep `test_datetime` at least. I can look into refactoring and removing duplicates after.

**Comment by timcolonel** [comment]
Also I was wondering why you where using dep instead of bundler as they will only be used for contributing and most people use bundler anyway.

**Comment by emancu** [comment]
I'm ok leaving `test_grammar`  for specific cases.

In the other hand, about `bundler` let me suggest you a blog post: http://bits.citrusbyte.com/managing-ruby-dependencies

> **tl;dr** Bundler is a very big dependency that solves a lot of problems that I don’t have

File: test/toml_test.rb [hunk 5]
```diff
@@ -1,7 +1,9 @@
 require_relative 'helper'
 require_relative 'toml_examples'
+require 'json'
+require 'pp'
```
Comment (emancu) [inline test/toml_test.rb]: Why is this needed?

Comment (timcolonel) [inline test/toml_test.rb]: i forgot to remove it i was trying to have a better compare view of the json and toml in case they differ but pp doesn't seem to improve much

**Comment by timcolonel** [comment]
Thanks for the article, I see your point.

**Comment by emancu** [comment]
Is it ready for merge ?

On Monday, June 8, 2015, Timothee Guerin <notifications@github.com> wrote:

> Thanks for the article, I see your point.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/emancu/toml-rb/pull/71#issuecomment-110078115>.
>


-- 
Emiliano Mancuso <http://about.me/emancu>

**Comment by timcolonel** [comment]
It should unless you need anything else
